### DISCUSSION TITLE
When will ASP.NET Core Implementation of Rate Limiter start?

### DISCUSSION THREAD

**Discussion by JAYARAMGITHUB** [ThreeMammals/Ocelot#2138] [Q&A] [answered]
I recently read about the potential redesign of Ocelot’s [Rate Limiting feature](https://ocelot.readthedocs.io/en/latest/features/ratelimiting.html) to utilize the .NET inbuilt `RateLimiter` class and the [System.Threading.RateLimiting](https://www.nuget.org/packages/System.Threading.RateLimiting) NuGet package, as discussed in the [Future and ASP.NET Core Implementation](https://ocelot.readthedocs.io/en/latest/features/ratelimiting.html#future-and-asp-net-core-implementation) article.

With .NET 8 now available and .NET 9 on the horizon, I’m eager to know when we might expect the new rate limiting feature using the .NET inbuilt Rate Limiter to be available in Ocelot. Any details on the roadmap or upcoming releases that will include this feature would be greatly appreciated.

Thank you for your continuous efforts in enhancing Ocelot. Looking forward to your response.

**Comment by raman-m** [comment]
Hello Jayaram,
We lack the necessary development capacity for this feature upgrade.
**As a team, we invite you to contribute!** Rest assured, we will offer guidance and conduct a detailed code review.

Concerning the RL [Roadmap](https://ocelot.readthedocs.io/en/develop/features/ratelimiting.html#roadmap)... The next [roadmap](https://ocelot.readthedocs.io/en/develop/features/ratelimiting.html#roadmap) will be published with next major release of version 24.x. Also, I will consult with the team, and I am committed to assisting in the development of a [roadmap](https://ocelot.readthedocs.io/en/develop/features/ratelimiting.html#roadmap) for this substantial upgrade, contingent upon your confirmed intention to contribute.
Could you share your experience with using ASP.NET's [native feature](https://learn.microsoft.com/en-us/aspnet/core/performance/rate-limit) in other applications? What are your expertise and skills regarding the native feature?

**Comment by gonghuilie** [comment]
I also hope that this function can be improved. When my gateway is deployed in clusters, Rate limit counter is not supported. I tried many methods, but failed.

**Reply by raman-m** [reply]
> I also hope that this function can be improved.

Certainly, we aim to prioritize the development of this feature once our current [backlog](https://github.com/ThreeMammals/Ocelot/issues) is cleared. We also welcome professional contributions at any time.

>When my gateway is **deployed in clusters**, Rate limit counter is not supported. I tried many methods, but failed.

The [answer](https://github.com/ThreeMammals/Ocelot/discussions/2098#discussioncomment-9773605) in discussion #2098 might be of assistance to you! Regrettably, the [DistributedCacheRateLimitStorage](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/RateLimiting/DistributedCacheRateLimitStorage.cs) feature remains unpublished in the [documentation](https://github.com/ThreeMammals/Ocelot/blob/develop/docs/features/ratelimiting.rst). We plan to [publish](https://github.com/ThreeMammals/Ocelot/issues/2141#issuecomment-2326054109) the feature shortly.

**Comment by yjorayev** [comment]
Hi,

I am planning to see if I can make changes to use .Net Rate Limiting middleware. I would like to also be able to configure to run rate limiter AFTER the authentication, so that auth context can be used for limiting(for ex: claims from token). 

Let me know your thoughts and I can start working on this soon.

**Reply by raman-m** [reply]
Apologies, Yhlas, are you asking a question or proposing a functionality?

**Reply by yjorayev** [reply]
Hi Raman, sorry for confusing message. I am asking 2 questions:

1. Is it okay for me to go ahead try integrate .Net Rate Limiting?
2. Do you have any thoughts/objections on allowing RateLimiter to run after Authentication?

**Reply by raman-m** [reply]
> 1. Is it okay for me to go ahead try integrate .Net Rate Limiting?

Certainly, integrating .NET Rate Limiting is a go!

However, please consider these challenges in our significant redesign:
- We need to maintain the existing rate limiting, also known as [Ocelot's Own Implementation](https://ocelot.readthedocs.io/en/latest/features/ratelimiting.html#ocelot-own-implementation). This feature of Ocelot is set to evolve in 2025, with plans to introduce rate limiting based on the client's IP address (currently, it's only by [header](https://ocelot.readthedocs.io/en/latest/features/ratelimiting.html#rate-limit-by-client-s-header), which has its limitations and security concerns).
- The integration of the [ASP.NET Implementation](https://ocelot.readthedocs.io/en/latest/features/ratelimiting.html#future-and-asp-net-implementation) should reuse current interfaces and, if necessary, expand the functionality with additional interfaces for DI.
- We must create JSON options that allow for an effortless switch between the two implementations.
- Ideally, we aim for zero breaking changes, though minor upgrade instructions would be acceptable.

**Are you up for this development challenge?**
How many years have you been coding in C# for your .NET projects? I've reviewed your profile and noticed TypeScript and Python projects, which suggests that C# might not be your primary skill, correct?

**Reply by raman-m** [reply]
> 2. Do you have any thoughts/objections on allowing RateLimiter to run after Authentication?

Apologies, I will respond to this question later, either tonight or tomorrow...

**Reply by yjorayev** [reply]
Thank you, I will keep challenges you mentioned in mind.

> 
> **Are you up for this development challenge?** How many years have you been coding in C# for your .NET projects? I've reviewed your profile and noticed TypeScript and Python projects, which suggests that C# might not be your primary skill, correct?

I do .Net development for work since 2017, so I'd say C# is my primary programming language. Most of work in my personal profile is from college years.

**Reply by raman-m** [reply]
- Official PR: #2188

**Reply by raman-m** [reply]
> 2. Do you have any thoughts/objections on allowing RateLimiter to run after Authentication?

Sure, I have concerns: current [RateLimitingMiddleware](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/RateLimiting/Middleware/RateLimitingMiddleware.cs) must work for anonymous traffic. So, the current disposition is &rarr;
https://github.com/ThreeMammals/Ocelot/blob/d310508aa28997a492eea4bc5cff889eb43014dd/src/Ocelot/Middleware/OcelotPipelineExtensions.cs#L79-L92
So, I'm just curious, why did you ask the question?
>  I would like to also be able to configure to run rate limiter AFTER the authentication, so that auth context can be used for limiting(for ex: claims from token).

I guess this must be a new feature and new middleware. After the Ocelot app starts, its pipeline is static and can’t be changed. However, we can manage this by adding new middleware after the authorization one. You should keep in mind that routes can be anonymous or authenticated, so we need two middlewares to manage this.

What similar analogs do you know in the .NET world? Does Microsoft ASP.NET rate limiter process authenticated traffic? How? Could you share some links, please?

**Comment by raman-m** [comment, marked as answer]
### The development has been started in
- PR #2188
